flocker/control/_persistence.py

```diff
@@ -163,15 +252,66 @@ def __init__(self, reactor, path):
     def startService(self):
         if not self._path.exists():
             self._path.makedirs()
-        self._config_path = self._path.child(b"current_configuration.v1.json")
-        if self._config_path.exists():
+        self.load_configuration()
+        MultiService.startService(self)
+        _LOG_STARTUP(configuration=self.get()).write(self.logger)
+
+    def _versioned_config(self, force_latest=False):
+        """
+        Sequentially and decrementally check for versioned configuration
+        files, from the current version to version 1. Return the file
+        found along with its version number, or if no valid
+        configuration file exists, return the current version number
+        and the expected file path of a configuration file matching
+        the latest version.
+
+        :return: A ``tuple`` comprising a version ``int`` and
+            config ``FilePath``.
+        """
+        config_files = [
+            (
+                version,
+                self._path.child(_VERSIONED_CONFIG_FILE % version)
+            )
+            for version in range(_CURRENT_VERSION, -1, -1)
+        ]
+        # Check for each possible versioned config file, from newest
+        # to oldest and return the first match.
+        for version, config_file in config_files:
+            if config_file.exists():
+                return (version, config_file)
+        return (
+            _CURRENT_VERSION,
+            self._path.child(
+                b"current_configuration.v%d.json" % _CURRENT_VERSION
+            )
+        )
+
+    def load_configuration(self):
+        """
+        Load the persisted configuration, upgrading the configuration format
+        if an older version is detected.
+        """
+        self._config_version, self._config_path = self._versioned_config()
```

Comment:
There's no reason to set these as instance attributes; they're temporary state. In general avoid mutating or modifying objects when it can be avoid; this will result in code that is simpler to reason about.